src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Ec2Metadata.scala

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright (c) 2015 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.scalatracker
+
+// Scala
+import scala.concurrent.{ Await, Future }
+import scala.concurrent.duration._
+
+// Spray
+import spray.http._
+import spray.client.pipelining._
+
+// json4s
+import org.json4s._
+import org.json4s.JsonDSL._
+import org.json4s.jackson.JsonMethods._
+
+// This library
+import emitters.RequestUtils
+
+/**
+ * Trait with parsing EC2 meta data logic
+ * http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html
+ */
+object Ec2Metadata {
+  import RequestUtils.system.dispatcher
+  val pipeline: HttpRequest => Future[HttpResponse] = RequestUtils.pipeline
+
+  val instanceIdentitySchema = "iglu:com.amazon.aws.ec2/instance_identity_document"
+  val instanceIdentityUri = "http://169.254.169.254/latest/dynamic/instance-identity/document/"
+
+  private var contextSlot: Option[SelfDescribingJson] = None
+
+  /**
+   * Get context stored in mutable variable
+   *
+   * @return some context or None in case of any error or not completed request
+   */
+  def context: Option[SelfDescribingJson] = contextSlot
+
+  /**
+   * Set callback on successful instance identity GET request
+   */
+  def initializeContextRequest: Unit = {
+    getInstanceContextFuture.onSuccess {
+      case json: SelfDescribingJson => contextSlot = Some(json)
+    }
+  }
+
+  /**
+   * Tries to make blocking request to EC2 instance identity document
+   * On EC2 request takes ~6ms, while on non-EC2 box it blocks thread for 3 second
+   *
+   * @return some context or None in case of any error including 3 sec timeout
+   */
+  def getInstanceContextBlocking: Option[SelfDescribingJson] =
+    try {
+      Some(Await.result(getInstanceContextFuture, 3 seconds))
+    }
+    catch {
+      case _: Throwable => None
```

Comment:
Can we use `scala.util.control.NonFatal` here so we don't accidentally catch any OutOfMemoryErrors?

https://tersesystems.com/2012/12/27/error-handling-in-scala/